os/kernel/binary_manager/binary_manager.c

```diff
@@ -145,6 +154,96 @@ void binary_manager_update_running_state(int bin_id)
 	binary_manager_notify_state_changed(bin_idx, BINARY_STARTED);
 }
 
+#ifdef CONFIG_BINMGR_RECOVERY
+static faultmsg_t *binary_manager_alloc_faultmsg(void)
+{
+	return (faultmsg_t *)sq_remfirst((sq_queue_t *)&g_freemsg_list);
+}
+
+static void binary_manager_free_faultmsg(faultmsg_t *msg)
+{
+	sq_addlast((FAR sq_entry_t *)msg, (FAR sq_queue_t *)&g_freemsg_list);
+}
+
+static int faultmsg_sender(int argc, char *argv[])
+{
+	int idx;
+	int ret;
+	faultmsg_t *msg;
+	binmgr_request_t request_msg;
+
+	/* Initialize pre-allocated fault messages */
+
+	for (idx = 0; idx < FAULTMSG_COUNT; idx++) {
+		sq_addlast((FAR sq_entry_t *)&g_prealloc_faultmsg[idx], (FAR sq_queue_t *)&g_freemsg_list);
+	}
+
+	while (1) {
+		/* Wait for fault messages and handle it */
+		up_block_task(this_task(), TSTATE_WAIT_UNBLOCK);
+		while (!sq_empty(&g_faultmsg_list)) {
+			msg = (struct faultmsg_s *)sq_remfirst(&g_faultmsg_list);
+			request_msg.cmd = BINMGR_FAULT;
+			request_msg.requester_pid = msg->binid;
+			bmllvdbg("Send fault message, bin id %d\n", request_msg.requester_pid);
+			ret = mq_send(g_binmgr_mq_fd, (const char *)&request_msg, sizeof(binmgr_request_t), BINMGR_FAULT_PRIO);
+			ASSERT(ret == OK);
+			binary_manager_free_faultmsg(msg);
```

Comment:
Some cases, you made a static function for single-queue functionality. But sometimes, you just call tha sq_XX directly. Is there any reason?